crates/client/src/client/rc_stream.rs

```diff
@@ -0,0 +1,122 @@
+// Copyright 2015-2016 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::pin::Pin;
+use std::sync::Arc;
+use std::task::{Context, Poll};
+
+use futures_util::lock::Mutex;
+use futures_util::stream::{Fuse, Stream, StreamExt};
+use futures_util::{ready, FutureExt};
+
+#[allow(clippy::type_complexity)]
+#[must_use = "stream do nothing unless polled"]
+pub(crate) struct RcStream<S: Stream>
+where
+    S: Stream + Send + Unpin,
+    S::Item: Clone + Send,
+{
+    stream_and_result: Arc<Mutex<(Fuse<S>, Vec<S::Item>)>>,
```

Comment:
I think we need some limit on this... I'm not sure what it should be though. Maybe we should change it to a `VecDeque` and when we hit the limit, pop at the front and append at the back? The semantics would be awkward if someone misses the front of the stream though.